allennlp/modules/span_extractors/locally_normalised_span_extractor.py

```diff
@@ -0,0 +1,96 @@
+import torch
+from overrides import overrides
+
+from allennlp.common.params import Params
+from allennlp.modules.span_extractors.span_extractor import SpanExtractor
+from allennlp.modules.time_distributed import TimeDistributed
+from allennlp.nn import util
+
+@SpanExtractor.register("locally_normalised")
+class LocallyNormalisedSpanExtractor(SpanExtractor):
+    """
+    Computes span representations by generating an unnormalized attention score for each
+    word in the document, compute distributions over every span with respect to these
+    scores by normalising the attention scores for words inside the span.
+
+    Given these attention distributions over every span, weight the corresponding vector
+    representations of the words in the span by this distribution, returning a weighted
+    representation of each span.
+
+    Parameters
+    ----------
+    input_dim : ``int``, required.
+        The last dimension of the sequence representations.
+
+    Returns
+    -------
+    attended_text_embeddings : ``torch.FloatTensor``
+        A tensor of shape (batch_size, num_spans, input_dim), which each span representation
+        is formed by locally normalising a global attention over the sequence. The only way
+        in which the attention distribution differs over different spans is in the set of words
+        over which they are normalized.
+    """
+    def __init__(self,
+                 input_dim: int) -> None:
+        super().__init__()
+        self._global_attention = TimeDistributed(torch.nn.Linear(input_dim, 1))
+
+    @overrides
+    def forward(self, # pylint: disable=arguments-differ
+                sequence_tensor: torch.FloatTensor,
+                indicies: torch.LongTensor) -> None:
+        # both of shape (batch_size, num_spans, 1)
+        span_starts, span_ends = indicies.split(1, dim=-1)
+
+        # shape (batch_size, num_spans, 1)
+        span_widths = span_ends - span_starts
+
+        # We need to know the maximum span width so we can
+        # generate indices to extract the spans from the sequence tensor.
+        # These indices will then get masked below, such that if the length
+        # of a given span is smaller than the max, the rest of the values
+        # are masked.
+        max_batch_span_width = int(span_widths.max().data) + 1
+
+        # shape (batch_size, sequence_length, 1)
+        global_attention_logits = self._global_attention(sequence_tensor)
+
+        # Shape: (1, 1, max_batch_span_width)
+        max_span_range_indices = util.get_range_vector(max_batch_span_width,
+                                                       sequence_tensor.is_cuda).view(1, 1, -1)
```

Comment:
Unrelated to this PR, but something I've been wondering: do you know if this `.is_cuda()` even works when you have multiple GPUs?  We're just calling `.cuda()` on the resulting tensor, without specifying a device, and this could be breaking things.  Should we be passing a tensor to this function to copy from instead of a boolean flag?